File: CS/T686902/EditableColumnHeaderBehavior.cs
using DevExpress.Mvvm.UI;
using DevExpress.Mvvm.UI.Interactivity;
using DevExpress.Xpf.Bars;
using DevExpress.Xpf.Editors;
using DevExpress.Xpf.Grid;
using System;
using System.Linq;
using System.Windows;
using System.Windows.Threading;

namespace T686902 {
    class EditableColumnHeaderBehavior : Behavior<GridControl> {
        public static bool GetAllowEditHeader(DependencyObject obj) {
            return (bool)obj.GetValue(AllowEditHeaderProperty);
        }
        public static void SetAllowEditHeader(DependencyObject obj, bool value) {
            obj.SetValue(AllowEditHeaderProperty, value);
        }
        public static readonly DependencyProperty AllowEditHeaderProperty = DependencyProperty.RegisterAttached("AllowEditHeader", typeof(bool), typeof(EditableColumnHeaderBehavior), new PropertyMetadata(false, OnAllowEditHeaderChanged));

        static void OnAllowEditHeaderChanged(DependencyObject d, DependencyPropertyChangedEventArgs e) {
            GridColumn column = d as GridColumn;
            if (column != null && column.Header == null && (bool)e.NewValue)
                column.Header = column.FieldName;
        }
        GridControl Grid => AssociatedObject;
        TableView View => (TableView)Grid.View;

        protected override void OnAttached() {
            base.OnAttached();
            View.ColumnHeaderTemplate = (DataTemplate)Grid.FindResource("EditableHeaderTemplate");
            View.ShowGridMenu += View_ShowGridMenu;
        }

        private void View_ShowGridMenu(object sender, GridMenuEventArgs e) {
            if (e.MenuType == GridMenuType.Column) {
                GridColumnHeader columnHeader = e.TargetElement as GridColumnHeader;
                GridColumn column = columnHeader.DataContext as GridColumn;
                bool allowEditColumnHeader = GetAllowEditHeader(column);
                BarButtonItem item = new BarButtonItem();
                item.Content = allowEditColumnHeader ? "Hide ColumnHeader Editor" : "Show ColumnHeader Editor";
                item.ItemClick += Item_ItemClick;
                item.Tag = new object[] { column, columnHeader };
                e.Customizations.Add(item);
            }
        }

        private void Item_ItemClick(object sender, ItemClickEventArgs e) {
            object[] objs = (object[])e.Item.Tag;
            GridColumn column = (GridColumn)objs[0];
            SetAllowEditHeader(column, !GetAllowEditHeader(column));
            if (GetAllowEditHeader(column)) {
                GridColumnHeader header = (GridColumnHeader)objs[1];
                TextEdit te = LayoutTreeHelper.GetVisualChildren(header).OfType<TextEdit>().FirstOrDefault();
                Dispatcher.BeginInvoke(new Action(() => {
                    te.Focus();
                    te.SelectAll();
                    te.LostFocus += Te_LostFocus;
                }), DispatcherPriority.Normal);
            }
        }

        private void Te_LostFocus(object sender, RoutedEventArgs e) {
            TextEdit te = (TextEdit)sender;
            te.Visibility = Visibility.Hidden;
            te.LostFocus -= Te_LostFocus;
        }

        protected override void OnDetaching() {
            base.OnDetaching();
            View.ShowGridMenu -= View_ShowGridMenu;
        }
    }
}

File: VB/T686902/EditableColumnHeaderBehavior.vb
Imports DevExpress.Mvvm.UI
Imports DevExpress.Mvvm.UI.Interactivity
Imports DevExpress.Xpf.Bars
Imports DevExpress.Xpf.Editors
Imports DevExpress.Xpf.Grid
Imports System
Imports System.Linq
Imports System.Windows
Imports System.Windows.Threading

Namespace T686902
    Friend Class EditableColumnHeaderBehavior
        Inherits Behavior(Of GridControl)

        Public Shared Function GetAllowEditHeader(ByVal obj As DependencyObject) As Boolean
            Return DirectCast(obj.GetValue(AllowEditHeaderProperty), Boolean)
        End Function
        Public Shared Sub SetAllowEditHeader(ByVal obj As DependencyObject, ByVal value As Boolean)
            obj.SetValue(AllowEditHeaderProperty, value)
        End Sub
        Public Shared ReadOnly AllowEditHeaderProperty As DependencyProperty = DependencyProperty.RegisterAttached("AllowEditHeader", GetType(Boolean), GetType(EditableColumnHeaderBehavior), New PropertyMetadata(False, AddressOf OnAllowEditHeaderChanged))

        Private Shared Sub OnAllowEditHeaderChanged(ByVal d As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
            Dim column As GridColumn = TryCast(d, GridColumn)
            If column IsNot Nothing AndAlso column.Header Is Nothing AndAlso DirectCast(e.NewValue, Boolean) Then
                column.Header = column.FieldName
            End If
        End Sub
        Private ReadOnly Property Grid() As GridControl
            Get
                Return AssociatedObject
            End Get
        End Property
        Private ReadOnly Property View() As TableView
            Get
                Return CType(Grid.View, TableView)
            End Get
        End Property

        Protected Overrides Sub OnAttached()
            MyBase.OnAttached()
            View.ColumnHeaderTemplate = CType(Grid.FindResource("EditableHeaderTemplate"), DataTemplate)
            AddHandler View.ShowGridMenu, AddressOf View_ShowGridMenu
        End Sub

        Private Sub View_ShowGridMenu(ByVal sender As Object, ByVal e As GridMenuEventArgs)
            If e.MenuType = GridMenuType.Column Then
                Dim columnHeader As GridColumnHeader = TryCast(e.TargetElement, GridColumnHeader)
                Dim column As GridColumn = TryCast(columnHeader.DataContext, GridColumn)
                Dim allowEditColumnHeader As Boolean = GetAllowEditHeader(column)
                Dim item As New BarButtonItem()
                item.Content = If(allowEditColumnHeader, "Hide ColumnHeader Editor", "Show ColumnHeader Editor")
                AddHandler item.ItemClick, AddressOf Item_ItemClick
                item.Tag = New Object() { column, columnHeader }
                e.Customizations.Add(item)
            End If
        End Sub

        Private Sub Item_ItemClick(ByVal sender As Object, ByVal e As ItemClickEventArgs)
            Dim objs() As Object = CType(e.Item.Tag, Object())
            Dim column As GridColumn = DirectCast(objs(0), GridColumn)
            SetAllowEditHeader(column, Not GetAllowEditHeader(column))
            If GetAllowEditHeader(column) Then
                Dim header As GridColumnHeader = DirectCast(objs(1), GridColumnHeader)
                Dim te As TextEdit = LayoutTreeHelper.GetVisualChildren(header).OfType(Of TextEdit)().FirstOrDefault()
                Dispatcher.BeginInvoke(New Action(Sub()
                    te.Focus()
                    te.SelectAll()
                    AddHandler te.LostFocus, AddressOf Te_LostFocus
                End Sub), DispatcherPriority.Normal)
            End If
        End Sub

        Private Sub Te_LostFocus(ByVal sender As Object, ByVal e As RoutedEventArgs)
            Dim te As TextEdit = DirectCast(sender, TextEdit)
            te.Visibility = Visibility.Hidden
            RemoveHandler te.LostFocus, AddressOf Te_LostFocus
        End Sub

        Protected Overrides Sub OnDetaching()
            MyBase.OnDetaching()
            RemoveHandler View.ShowGridMenu, AddressOf View_ShowGridMenu
        End Sub
    End Class
End Namespace
